tool-openssl/x509_test.cc

```diff
@@ -47,59 +48,336 @@ X509* CreateAndSignX509Certificate() {
+
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509(in_file.get(), x509.get()));
+  }
+
+  args_list_t args = {"-in", in_path, "-out", out_path, "-signkey", signkey_path, "-days", "365"};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+  RemoveFile(out_path);
+  RemoveFile(signkey_path);
+}
+
+// Test -dates
+TEST(X509Test, X509ToolDatesTest) {
+  char in_path[PATH_MAX];
+
+  ASSERT_GT(createTempFILEpath(in_path), 0u);
+
+  bssl::UniquePtr<X509> x509(CreateAndSignX509Certificate());
+  ASSERT_TRUE(x509);
 
-    remove(in_path);
-    remove(out_path);
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509(in_file.get(), x509.get()));
+  }
+
+  args_list_t args = {"-in", in_path, "-dates"};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+}
+
+// Test -req
+TEST(X509Test, X509ToolReqTest) {
+  char in_path[PATH_MAX];
+  char out_path[PATH_MAX];
+  char signkey_path[PATH_MAX];
+
+  ASSERT_GT(createTempFILEpath(in_path), 0u);
+  ASSERT_GT(createTempFILEpath(out_path), 0u);
+  ASSERT_GT(createTempFILEpath(signkey_path), 0u);
+
+  bssl::UniquePtr<EVP_PKEY> pkey(EVP_PKEY_new());
+  ASSERT_TRUE(pkey);
+
+  bssl::UniquePtr<RSA> rsa(RSA_new());
+  ASSERT_TRUE(rsa);
+  bssl::UniquePtr<BIGNUM> bn(BN_new());
+  ASSERT_TRUE(bn && BN_set_word(bn.get(), RSA_F4) && RSA_generate_key_ex(rsa.get(), 2048, bn.get(), nullptr));
+  ASSERT_TRUE(EVP_PKEY_assign_RSA(pkey.get(), rsa.release()));
+
+  {
+    ScopedFILE signkey_file(fopen(signkey_path, "wb"));
+    ASSERT_TRUE(signkey_file);
+    ASSERT_TRUE(PEM_write_PrivateKey(signkey_file.get(), pkey.get(), nullptr, nullptr, 0, nullptr, nullptr));
+  }
+
+  bssl::UniquePtr<X509_REQ> req(X509_REQ_new());
+  ASSERT_TRUE(req);
+  X509_REQ_set_pubkey(req.get(), pkey.get());
+  X509_REQ_sign(req.get(), pkey.get(), EVP_sha256());
+
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509_REQ(in_file.get(), req.get()));
+  }
+
+  args_list_t args = {"-in", in_path, "-out", out_path, "-req", "-signkey", signkey_path};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+  RemoveFile(out_path);
+  RemoveFile(signkey_path);
+}
+
+// Test -checkend
+TEST(X509Test, X509ToolCheckEndTest) {
+  char in_path[PATH_MAX];
+
+  ASSERT_GT(createTempFILEpath(in_path), 0u);
+
+  bssl::UniquePtr<X509> x509(CreateAndSignX509Certificate());
+  ASSERT_TRUE(x509);
+
+  {
+    ScopedFILE in_file(fopen(in_path, "wb"));
+    ASSERT_TRUE(in_file);
+    ASSERT_TRUE(PEM_write_X509(in_file.get(), x509.get()));
+  }
+
+  args_list_t args = {"-in", in_path, "-checkend", "3600"};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+}
+
+// Test mutually exclusive options, required options, and required arugments
+TEST(X509Test, MutuallyExclusiveOptionsTest) {
```

Comment:
Great test! It does seems like we can parameterize the `args` options here with `TEST_P` and C++ classes. It's just a potential improvement, but not really a priority. This might serve as a good example:

https://github.com/aws/aws-lc/blob/1fbf58437a32da4008774eed92530b787d210b1e/crypto/cipher_extra/cipher_test.cc#L591-L616

Reply:
Currently trying to implement but errors preventing build. Will come back to this later